doc/docs/spec/typeclasses/index.md

```diff
@@ -0,0 +1,124 @@
+# Typeclasses
+
+Typeclasses are a feature that comes from Haskell as a means of achieving overloading of sorts. In Vapor this is different,
+as the main goal to achieve with typeclasses is allowing users to customize behavior, but also in more ways than once. This
+document tries to specify how they look like and how they behave.
+
+## Syntax
+
+The typeclass support encompasses the following keywords:
+
+ * `typeclass`, used as a kind for the declaration of typeclasses.
+ * `implicit`, which signifies that a given typeclass is - by design - to be inferred for all types, if no instances are
+   defined. **Drafting node: need an opt-out mechanism.**
+ * `instance`, used to declare an instance of a typeclass.
+ * `default`, which marks the instance that is used when no instance is explicitly specified.
+ * `with`, which will ultimately be used properly for introducing all kinds of templates, but right now its usage is
+   confined to defining typeclasses, as templates are not a feature to fully land in the minimal viable product.
+
+Similarly to structures, typeclasses can be defined in the named an unnamed forms; the named one is translated to a declaration
+of an entity initialized with the unnamed form of the definition. The grammar is as follows (this duplicates some terms that
+will be defined elsewhere, but not all of them, since quoting the entire grammar of expressions here would be excessive):
+
+```
+named-typeclass-declaration ::= template-introducer `implicit`? `typeclass` identifier
+    `{` typeclass-function* `}` `;`
+typeclass-literal ::= template-introducer `implicit`? `typeclass` `{` function-declaration* `}`
+
+template-introducer ::= `with` `(` parameter (`,` parameter)* `)`
+parameter ::= identifier `:` expression
+
+typeclass-function ::= function-declaration | function-definition
+function-declaration ::= function-signature `;`
+function-definition ::= function-signature block
+function-signature ::= `function` identifier `(` (parameter (`,` parameter)*)? `)` (`->` trailing-type)?
+```
+
+A definition of a default instance is itself a declaration and cannot be used as an expression. Other definitions of
+typeclass instances are expressions and can be used as such.
+
+
+```
+default-instance-definition ::= `default` `instance` identifier `(` expression `)`
+    `{` function-definition* `}` `;`
+instance-literal ::= `instance` identifier `(` expression `)` `{` function-definition* `}`
+```
+
+## Semantics
+
+This section uses names from the following piece of code to define semantics and requirements:
+
+```
+let tc = with (T : type) typeclass {        // 1
```

Comment:
Only shown here is `T : type`, but that implies, to me, that `T : C` is valid. Some examples:

### Integer
```
let bits = with (Bits : int) typeclass {
};
```

### Typeclasses
```
let bits = with (Bits : int) typeclass {
};

let bits32 = with (Bits : bits(32)) typeclass {
};
```

### More complex usage
```
let render = with (Dimension : int) typeclass {
    function point(v : vector(Dimension)) -> void;
};
```

Are all of these valid? If so, I think it'd be worth adding a note and/or an example of a non-type parameter to a typeclass.

Reply:
I don't think I want value *type*classes to be valid. C++ is trying to do something with value concepts, but I don't think this is reasonable in the model I'm thinking about. This is explicitly meant to be used as a customization point mechanism, which makes no sense for values... I think.

At one point in the future Vapor will have a sort of "template template parameters", then those should be usable as another kind of an argument here. Simiarly you could specify a thing requiring a typeclass to be fulfilled here, but I don't see how values are useful.